README.md
bsdsum
======

Last modified: 2023/01/31.

  bsdsum is a tool for computing and checking digests of data. It runs under
GNU/Linux and BSD systems. Its name comes from the fact it is BSD-licensed and
derived from the BSD 'cksum' program. The tool supports the following digests
formats:

  - BSD format: "ALG (FILE) = DIGEST_HEX_ENCODED", or,
                "ALG (FILE) = DIGEST_BASE64_ENCODED", and/or
                "SIZE (FILE) = size_of_file"

  - GNU coreutils format: "DIGEST  FILE" (only md5, sha1, sha2)

  - cksum variant: "DIGEST FILE" (one space, only md5, sha1, sha2)

  The style of output is specified with the "-s" option:

    -s default: for BSD format (default style if unspecified)
    -s base64: for BSD base64-encoded format
    -s mix32: BSD format using a custom character set including 32 punctuation
              signs, letters, digits
    -s gnu: for GNU coreutils output
    -s cksum: for cksum variant
    -s terse: output the raw hexadecimal result without the file name
    -s binary: output the raw binary digest

  Supported algorithms are:

   MD5, SHA1, SHA256 (default choice if not specified), SHA384, SHA512,
   Keccak-1600 SHA3-256 and SHA3-512,
   WHIRLPOOL (512 bits),
   BLAKE224, BLAKE256, BLAKE384, BLAKE512,
   BLAKE2B (512 bits, unkeyed), BLAKE2S (256 bits, unkeyed),
   BLAKE3 (256 bits, unkeyed),
   SIZE (output the length of the file). 

  Multiple algorithms may be specified for each source of data, using the -a 
option (i.e. -a md5,sha1,size).

  On Linux (both 32-bit and 64-bit) and 64-bit BSDs, large files support is
available, else the maximum length of the files is 2 or 4 GB and block devices
are not supported.


Split digests
=============

  In addition, a special "split" mode can be used to compute digests of files
only (not data sent thru stdin). The split mode is enabled using "-a ALG:N",
where ALG is one of:

  MD5, SHAxxx, SHA3-xxx, WHIRLPOOL, BLAKExxx

  and N an integer greater or equal than 2. 

  Each source file is split into N equal parts and each part is hashed in a 
separate thread to produce the digest D(i). Once the digests D(i) i=1..N are 
computed, they are concatenated and the final digest is the hash of this data:
ALG(D(1)|...|D(N)). This let you use all the cpus of your machine; with two 
cpus, SHA256:2 is two times faster than SHA256.

  *note* that the digests produced using ALG and ALG:N are NOT the same.


Usage
=====

  - read the embedded help:

        bsdsum -h

  - compute and display the SHA256 digest of file 'data':

        bsdsum -a sha256 data

    or

        cat data | bsdsum -a sha256

  - store the SHA256:2 split digest of file 'data' into 'data.dg'
    (the 'cat' method above will not work for split digests, as
    the length of the input data must be known):

        bsdsum -a sha256:2 -o data.dg data

    or

        bsdsum -a sha256:2 data > data.dg

  - fully check the digests stored in the file 'data.dg' (multiple lists to
    check can be specified on the command-line):

        bsdsum -c data.dg

  - fully check the list of digests read from standard input:

        bsdsum -c

  - check the digest for file 'my_file' from the list of digests 'data.dg'
    (only one list of files to check can be specified when using -C):

        bsdsum -C data.dg my_file

    Note: for such a query, an error is reported if 'my_file' is not found in
          the list 'data.dg'.

  - check the MD5 digest (only) for file 'my_file' from the list of digests
    'data.dg':

        bsdsum -C data.dg -a md5 my_file

  - store the SHA3-256 digest, base64-encoded and the length of file 'data'
    into 'data.dg':

        bsdsum -s base64 -a sha3-256,size data > data.dg

  - compute the SHA-256 digest of the data in file './data' from offset 100
    to 149 (only one file to digest must be specified here, and the output
    is always done in "terse" style):

        bsdsum -f 100 -l 50 ./data

  - hash any file under directory ./src/ (recursively), but skipping links:

        bsdsum -r -k ./src

  - run the tiny auto-test:

        bsdsum -t


Build & install
===============

* tested systems: GNU/Linux, FreeBSD, OpenBSD, NetBSD, DragonFlyBSD, using
  either gcc or clang C compiler

* dependencies: libc only

* first run the "./configure" script. Then, on success, run "make" to build
  the binary out/bsdsum and run some tests. You can also run "make static"
  to produce the static binary "out/bsdsum-static" that has no dependency.

* you should now run "make check" to launch the tiny test suite

* install the files using "make install" (under root user). You can add
  DESTDIR=xxx to install files under the directory xxx/ (by default, equal to
  the prefix setup by the configure script).

* configure script options:
  --prefix=...   specify installation/configuration prefix (default /usr/local)
  --cc=...   to specify the name of the C compiler (gcc, clang, default: auto)

* other make targets:
  make static: build out/bsdsum-static (static binary)
  make clean: clean building tree, keeping configuration
  make distclean: clear all generated files
  make check: run the tests


Lists of digests
================

  A list of digests is a file (as produced using the -o option) that stores
hashes for many other files. Such a file may be specified when using the
-c/-C options to verify a list of digests. Valid lines are:

  - empty ones (ignored),

  - lines starting with # (comments, ignored)

  - "ALG(FILE) = DIGEST"

  - "DIGEST  FILE" (two spaces)

  - "DIGEST FILE" (one space)

  - "# $BSDSUM: date,digest $" as last line of the file, stores the "supersum"
    of the list, that is, the digest of all valid lines (except ignored ones)
    above itself.

  Any other line will be rejected and an error reported.




ChangeLog
bsdsum
======

see git log for more details.

2023/01 - v1.5
* add BLAKE, BLAKE2B, BLAKE2S, BLAKE3, WHIRLPOOL algorithms
* add "mix32" (5-bit) encoding 
* add "-r" option for recursive directory access
* add "-k" option for skipping links
* add "-v" option for message reporting settings
* fix "-l" (length) and "-f" (offset) options
* add large files support on Linux (32-bit or 64-bit)
* add block devices support (only available if large files
  support is enabled)
* code reorganization
* many fixes

2022/12, 2023/01 - v1.4
* add split digest
* add binary style
* add configure script
* rewrite filelist parser
* add gcc or clang support
* add tests
* fix BSD support

2022/12/18 - v1.3
* use mmap to access content of files
* add -f (offset) and -l (length) options

2022/12/17 - v1.2
* add SHA3-256 and SHA3-512 support
* add again auto-test option [-t]
* add -s STYLE option
* change -h to help and old -h to -o
* change various options and flags

2022/12/11 - v1.1
* add libcrypto files from libressl
* add -g option (GNU output format)
* remove some options



src/bsdsum_op.c
/*
 * Copyright (c) 2022-2023
 *      Nicolas Provost <dev AT npsoft DOT fr>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/mman.h>
#include <sys/wait.h>
#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>
#include <signal.h>

#include "bsdsum.h"

bsdsum_op_t bsdsum_ops[] = {
	{
		"SIZE", 16, STYLE_TXT | STYLE_NOSPLIT | STYLE_FIXED,
		(op_init_t)bsdsum_size_init,
		(op_update_t)bsdsum_size_update,
		(op_final_t)bsdsum_size_final,
	},
	{
		"MD5", MD5_DIGEST_LENGTH, STYLE_ANY,
		(op_init_t)MD5_Init,
		(op_update_t)MD5_Update,
		(op_final_t)MD5_Final,
	},
	{
		"SHA1", SHA_DIGEST_LENGTH, STYLE_ANY,
		(op_init_t)SHA1_Init,
		(op_update_t)SHA1_Update,
		(op_final_t)SHA1_Final,
	},
	{
		"SHA256", SHA256_DIGEST_LENGTH, STYLE_ANY,
		(op_init_t)SHA256_Init,
		(op_update_t)SHA256_Update,
		(op_final_t)SHA256_Final,
	},
	{
		"SHA384", SHA384_DIGEST_LENGTH, STYLE_ANY,
		(op_init_t)SHA384_Init,
		(op_update_t)SHA384_Update,
		(op_final_t)SHA384_Final,
	},
	{
		"SHA512", SHA512_DIGEST_LENGTH, STYLE_ANY,
		(op_init_t)SHA512_Init,
		(op_update_t)SHA512_Update,
		(op_final_t)SHA512_Final,
	},
	{
		"SHA3-256", SHA256_DIGEST_LENGTH, STYLE_BSD,
		(op_init_t)bsdsum_sha3_256_begin,
		(op_update_t)bsdsum_sha3_update,
		(op_final_t)bsdsum_sha3_final,
	},
	{
		"SHA3-512", SHA512_DIGEST_LENGTH, STYLE_BSD,
		(op_init_t)bsdsum_sha3_512_begin,
		(op_update_t)bsdsum_sha3_update,
		(op_final_t)bsdsum_sha3_final,
	},
	{
		"WHIRLPOOL", WHIRLPOOL_DIGEST_LEN, STYLE_BSD,
		(op_init_t)rhash_whirlpool_init,
		(op_update_t)rhash_whirlpool_update,
		(op_final_t)rhash_whirlpool_final,
	},
	{
		"BLAKE224", BLAKE224_DIGEST_LEN, STYLE_BSD,
		(op_init_t)blake224_init,
		(op_update_t)blake224_update,
		(op_final_t)blake224_final,
	},
	{
		"BLAKE256", BLAKE256_DIGEST_LEN, STYLE_BSD,
		(op_init_t)blake256_init,
		(op_update_t)blake256_update,
		(op_final_t)blake256_final,
	},
	{
		"BLAKE384", BLAKE384_DIGEST_LEN, STYLE_BSD,
		(op_init_t)blake384_init,
		(op_update_t)blake384_update,
		(op_final_t)blake384_final,
	},
	{
		"BLAKE512", BLAKE512_DIGEST_LEN, STYLE_BSD,
		(op_init_t)blake512_init,
		(op_update_t)blake512_update,
		(op_final_t)blake512_final,
	},
	{
		"BLAKE3", BLAKE3_OUT_LEN, STYLE_BSD,
		(op_init_t)blake3_hasher_init,
		(op_update_t)blake3_hasher_update,
		(op_final_t)blake3_final,
	},
	{
		"BLAKE2B", BLAKE2B_OUTBYTES, STYLE_BSD,
		(op_init_t)blake2b_start,
		(op_update_t)blake2b_update,
		(op_final_t)blake2b_end,
	},
	{
		"BLAKE2S", BLAKE2S_OUTBYTES, STYLE_BSD,
		(op_init_t)blake2s_start,
		(op_update_t)blake2s_update,
		(op_final_t)blake2s_end,
	},
	{
		NULL, 0, 0, NULL,
	}
};

bsdsum_op_t* bsdsum_op_get (const char* name)
{
	int i;

	for (i = 0; bsdsum_ops[i].name; i++)
	{
		if (strcasecmp (bsdsum_ops[i].name, name) == 0)
			return &bsdsum_ops[i];
	}
	return NULL;
}

/* Parse one algorithm name 'cp'. Returns NULL on error. 
 * If 'style' is not STYLE_NONE, check that the algorithm
 * that was found supports this style of output. */
bsdsum_op_t* bsdsum_op_find_alg (const char *cp, 
					bsdsum_style_t style,
					int quiet)
{
	bsdsum_op_t *hf;
	char *p;
	char *endptr;
	long l = 0;

	p = strchr(cp, ':');
	if (p) {
		*p++ = '\0';
		l = strtol(p, &endptr, 10);
		if ((endptr && *endptr) || 
			(l <= 1) || (l > MAX_SPLIT)) {
			bsdsum_log(LL_WARN,  "bad algorithm name \"%s:%s\"", cp, p);
			return NULL;
		}
	}

	hf = bsdsum_op_get(cp);
	if (hf == NULL || hf->name == NULL) {
		if ( ! quiet)
			bsdsum_log(LL_WARN, "unknown algorithm \"%s\"", cp);
		return NULL;
	}
	if ((hf->use_style & STYLE_NOSPLIT) && l) {
		if ( ! quiet)
			bsdsum_log(LL_WARN, "algorithm \"%s\" does not support split", cp);
		return NULL;
	}
	if ((style != STYLE_NONE) && ! (hf->use_style & STYLE_FIXED)) {
		if ((hf->use_style & style) != style) {
			if ( ! quiet)
				bsdsum_log(LL_WARN, "%s doesn't support given style", 
					hf->name);
			return NULL;
		}
	}
	hf->split = (int) l;
	return hf;
}

/* Returns true if the given style use a case sensitive
 * comparison. */
bool bsdsum_op_case_sensitive(bsdsum_style_t st)
{
	if (st & STYLE_B64)
		return true;

	return false;
}

/* find the first function having STYLE_GNU or STYLE_CKSUM flag 
 * whose output length is 'len'. */
bsdsum_op_t* bsdsum_op_for_length(size_t len)
{
	int i;

	for (i = 0; bsdsum_ops[i].name; i++) {
		if ( ! (bsdsum_ops[i].use_style & 
				(STYLE_CKSUM | STYLE_GNU)))
			continue;
		if (len == bsdsum_ops[i].digestlen * 2)
			return &bsdsum_ops[i];
	}
	return NULL;
}

